[Week8] API 문서 만들기, Docker image 만들기

yongjunleeme opened the pull request:
마지막주도 잘 부탁드립니다 !

yongjunleeme committed:
상품 서비스에 주석을 작성하라

상품 서비스의 역할을 설명했습니다.

yongjunleeme committed:
상품 컨트롤러의 역할을 설명하라

yongjunleeme committed:
세션 컨트롤러의 역할을 설명하라

yongjunleeme committed:
회원 컨트롤러의 역할을 설명하라

yongjunleeme committed:
인증 서비스의 역할을 설명하라

yongjunleeme committed:
'*'를 사용한 import문을 구체적 import문으로 변경하라

yongjunleeme committed:
회원 서비스의 역할을 설명하라

yongjunleeme committed:
Jwt 활용 클래스를 설명하라

yongjunleeme committed:
도메인의 역할을 설명하라

yongjunleeme committed:
API 문서 작성을 위한 의존성을 추가하라

- 마크업 문서로 변경해주는 asciidoctor를 추가합니다.
- API 문서를 만들어주는 restdocs를 추가합니다.

yongjunleeme committed:
테스트보다 asscidoctor를 먼저 수행할 설정을 추가하라

yongjunleeme committed:
버전관리가 필요없는 폴더를 추가하라

yongjunleeme committed:
자바 버전을 11로 변경하라

`List.of`를 사용하는 코드에서 심볼을 찾을 수 없다는 에러가 나왔습니다.
Java 9에서 `List.of`가 Java 9에 추가된 것으로 확인해 버전을
변경했습니다.

yongjunleeme committed:
상품 관련 Rest Docs를 작성하라

yongjunleeme commented:
Asciidoc에도 혹시 함수처럼 반복되는 문구를 캡슐화할 수 있는 방법이 있는지 여쭈어 보아도 될까요~? 더불어 Asciidoc을 보통 회사에서 사용하시는지 궁금합니다. 문법을 익히는 게 가성비(?)가 맞는 것인가 의문이 들어 질문을 남겨 놓습니다~

ssisksl77 commented on app/build.gradle:
```diff
@@ -18,7 +18,7 @@ plugins {
 }
 
-sourceCompatibility = '1.8'
+sourceCompatibility = '11'
```
좋습니다!

ssisksl77 commented:
1. include 로 간단한 보일러플레이트는 재사용할 수 있을 것 같습니다.
2. 사용하는 곳도 겪었고 사용하지 않는 곳도 겪었어보았어요! 아마 팀의 성향에 따라 다를 것 같아요! asciidoc을 단순한 마크다운이라고 생각하시면 그리 어렵진 않을꺼에요~ 팀에서 약속한 형태만 계속 사용할테니까요!

ssisksl77 commented on app/build.gradle:
```diff
@@ -78,6 +81,10 @@ dependencies {
+
+    // Spring REST Docs
+    asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor'
+    testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
```
이것을 추가하면 구체적으로 무엇을 사용할 수 있게 될까요?
'API 문서를 만들어주는' 이라는 것이 '마크업 문서로 변경해주는' 과 어떻게 다른가요?!

yongjunleeme replied:
헛 알아보지 않고 유추해서 쓰는 바람에 잘못 적은 것 같습니다;;; -..-;;;

## asciidoctor
블록 매크로 명령어를 사용할 수 있도록 해준다.  `*.adoc` 파일에서 `snippets`를 사용하면 `build/generated-snippets.`를 가리킬 수 있도록 자동으로 설정한다. 

## testImplementation
테스트 구현 설정에 `spring-restdocs-mockmvc` 의존성을 추가한다. MVC 테스트를 하려면 `MockMvc`를 사용한다.
테스트 구동 작업으로 요청과 응답에 관한 문서를 만들 수 있다.

출처: 
- https://docs.spring.io/spring-restdocs/docs/current/reference/html5/
- https://www.baeldung.com/spring-rest-docs

ssisksl77 replied:
아주 좋아요! 하지만 여기서 궁금한 점이 생겼어요!
REST Docs 를 쓸 때는 왠지 항상 MockMvc 객체를 쓸 것 같은데 왜 애초에 처음부터 spring-test 에 넣어놓지 않았을까요?

왠지 REST Docs 을 만들기 위한 테스트 프로젝트가 한 개가 아닌 것 같아요!

다른 것들을 찾아보고 어떤 경우에 사용하는지 알아보아요!

yongjunleeme replied:
- MockMvc: 단위테스트에 사용. 실제 서버 환경과 동일한 `@SpringBootTest`가 필요없어 `@WebMvcTest`로 Presentation Layer 빈들만 불러서 테스트한다.
- RestAssured : 웹 서비스를 검증하기 위해 end-to-end 테스트에 사용. `@SpringBootTest`로 실제 요청을 보내 테스트하며 실제 요청에 필요한 다양한 메서드를 제공
- WebTestClient: 스프링5에 도입된 리액티브(non-blocking) 테스트 방법

테스트 용도에 따라 다른 프로젝트를 사용하기 위해서 고정이 아닌 선택사항으로 둔 게 아닌가 싶습니다!

## 출처

- https://docs.spring.io/spring-restdocs/docs/current/reference/html5/
- https://tecoble.techcourse.co.kr/post/2020-08-19-rest-assured-vs-mock-mvc/

yongjunleeme committed:
Fix typo

yongjunleeme committed:
문서에서 path 파라미터를 사용하기 위해 `RestDocumentationRequestBuilders`를 추가하라

generated-sippets/get-prouct에 response-fields.adocs가 생성되지
않아 오류가 발생했습니다.

pathParameters를 사용하기 위한 요청을 위해
RestDocumentationRequestBuilders 메소드를 사용했습니다.

[공식문서](https://docs.spring.io/spring-restdocs/docs/1.0.0.BUILD-SNAPSHOT/reference/html5/#documenting-your-api-path-parameters)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -29,7 +33,16 @@ public AuthenticationService(UserRepository userRepository,
     }
 
-    public String login(String email, String password) {
+    /**
+     * 주어진 이메일과 비밀번호로 로그인하고 생성한 액세스 토큰을 리턴한다.
+     *
+     * @param email    이메일
+     * @param password 비밀번호
+     * @return 생성한 액세스 토큰
+     * @throws LoginFailException 로그인에 실패한 경우
+     */
```
이 메소드에서 발생할 일은 희박하지만 예로써 이용하겠습니다! ^^

만약에 인자가 N개가 추가된다고 하면
```
주어진 이메일과 비밀번호와 param1와 param2와 ...paramN 으로 로그인하고 생성한 엑세스 토큰을 리턴한다.
@param email 이메일
...
@param paramN 파라미터N
...
```

이렇게 문서가 변경되겠죠! 

그러므로 맨 위에 있는 개요에는 `주어진 이메일과 비밀번호` 는 너무 구체적인 설명인 것 같아요!
저라면 현재 두 가지를 고려해볼 것 같아요!

1. 해당 객체를 위한 인자용 객체를 만든다
2. 문서에서 이메일과 비밀번호를 없애고 "인자"라는 단어로 추상화한다.

yongjunleeme replied:
아하, 변경 가능성이 있다면 문서에서 추상화 단계를 높인 단어를 사용하란 말씀하신 것이지요? "인자"는 혹시 너무 추상적이진 않은가요~? 이메일도 인자, 비밀번호도 인자라면 구별이 안 되지는 않을까요~? 잘 모르지만 의문을 남겨 봅니다.

그리고 인자용 객체는 무엇인지, 어떻게 만들어야 할 지 궁금합니다!

yongjunleeme committed:
변경 가능성에 대비하여 추상적 단어로 변경하라

yongjunleeme commented:
오늘 마리아 DB 연결, 도커 배포를 시도해보려고 했는데 강의 영상 버퍼링이 전진을 막고 있습니다 ㅠ.ㅠ
공식적으로 오늘이 의견을 여쭐 마지막 기회일 것 같은데.. 기회를 활용해 질문 몇 가지 남겨 놓습니다!

- 혹시 이 도커 배포 이후 CI, CD는 어떻게 학습하면 좋을지 여쭈어도 될까요~?
- 취업이 목표라면 주워 듣기로는 여러 기술을 쓰는 것보다 한 기술을 심도있게 다룬 티가 나야한다던데.. 그렇다면 스프링의 어떤 분야를 깊게 파는 게 도움이 될지, 어떤 방법을 쓰면 좋을지 여쭤보고 싶습니다!
- 기술면접 등 취업 과정 준비를 위한 팁, 또는 더 미래에는 함께 일하고 싶은 개발자가 되려면 갖춰야 할 덕목 등의 조언도 주신다면 큰 도움이 될 것 같습니다!!

ssisksl77 commented:
CI,CD는 방법이 아주 많습니다. 회사에서 사용하는 툴을 먼저 익혀보는 것이 어떨까요? 하나를 알면 다른 툴을 사용할 때도 '내가 쓰던기능은 이 툴에선 어떻게 할 수 있지?' 라는 의문으로 시작하면 어떨까요? 젠킨스를 쓰고 있다면 젠킨스를 github action을 쓰고 있다면 github action을! 없다면 새로운 툴을 도입해보는거죠! 아니면 원하는 회사에서 사용하는 툴을 알아보고 익혀보는거에요!

취업은 항상 어려운 문제네요. 저는 마음가는 것을 익히는 것이 좋다고 생각해요. 시험공부처럼 배우고 무언가를 준비하고 취업을 해본적이 없어서 종립님이나 다른분들에게 물어보시는 것이 더 현실적인 조언을 드릴 수 있을 것 같아요. 

---

저는 사실 Java/Spring 공부를 조금하다가 Lisp에 빠져서 Lisp공부만 했습니다. Lisp를 사용하는 회사에 취직하겠다고 취업전선에 뛰어들었습니다. 하지만 관련 회사가 전혀 없다는 것을 알아차리고 Java/Spring을 쓰는 작은 회사에 들어가게 됩니다. 

여러회사를 옮기다가 컬리까지 오게되고 리뷰어로 계시는 종립님을 만났었죠. 그러다가 그린랩스라는 회사에서 Lisp계열 언어인 Clojure라는 언어를 사용한다는 소식을 듣고 지원하여 합류하게 됩니다. 빈 프로젝트 폴더에서 시작해서 하나의 프로젝트가 런칭될 때, 그 언어가 오로지 Lisp로 만들어졌을 때 저는 희열을 느꼈습니다.

세상에는 여러가지 개발자가 있습니다. 꼭 면접을 위해 무엇을 공부해야 하는가? 라기보다는 내가 무엇을 하고싶은가를 생각하면서 하나의 프로젝트를 만들어보는 것은 어떨까 라는 생각이 듭니다.

그럼에도 스프링의 어떤 분야를 깊게 파는게 도움이 될지! 라는 생각이 드신다면! Spring framework 코어, Spring MVC, Spring Data, Test 정도가 되겠네요!

참고로 저는 기술면접을 위한 공부는 한 적이 없습니다. 항상 현 회사에서 사용하는 기술에만 집중했고, 제가 관심있는 기술만 습득했어요. 이렇게 공부를 하면 기술면접을 볼 때, 서로가 관심있어하는 기술이 전혀 다르다는 것을 느낄 때도 있습니다. 여기서 그 회사에 가기위해 공부하는가, 아니면 나의 길을 가는가 선택을 하면 됩니다! 일단! 면접을 여러번 보면 무엇이 필요한지 느끼실 거에요! 저는 위에 설명했듯이 반골기질이 있어서 제가 관심있는 공부만 했었습니다. 그럼에도 운명처럼 만나는 회사는 있게 마련이더라고요. ㅎㅎ

미래에 함께 일하고 싶은 개발자라... 마치 누구에게 인문한적 조언을 하는 것같아서 감히 제가 어떤 조언도 할 수 없을 것 같아요! 하지만 한가지 분명한건 개발자라고 함께 일하고 싶은 사람이 타분야와 다르지는 않을거에요!

yongjunleeme commented:
와.. 개인적 이야기까지 꺼내서 해주신 고무적인 조언 감사드립니다! Lisp이라는 게 얼마나 재밌으셨는지 그 느낌이 궁금해지네요. 저는 Spring이 재밌다는 생각을 못해봤는데 그냥 하는 게 아니라 취업의 도구로 삼아서 그런 것 같습니다. 즐거움의 접근 포인트를 찾아보는 것도 의미있는 일이 되겠군요!

- 내가 하고 싶은 사이드 프로젝트를 통해 찾는 내가 좋아하는 기술
- 많은 면접을 통해 발견하는 내가 좋아하는 기술에 관심을 보이는 회사
- 스프링의 우선순위는 Spring Core, Spring MVC, Spring Data, Test
- 개발자도 회사원이다. 함께 일하고 싶은 회사원이 어떤 모습일까.

이번 조언뿐만 아니라 여태껏 해주신 코멘트들까지 하나하나 소화해보겠습니다! 그간 감사했습니다!!

yongjunleeme commented:
[8주차 회고](https://github.com/yongjunleeme/spring-code-soom/blob/main/docs/week8.MD) 남겨 놓습니다~!
